bodo/pandas/utils.py

```diff
@@ -824,6 +762,79 @@ def arrow_to_empty_df(arrow_schema):
     return _reconstruct_pandas_index(empty_df, arrow_schema)
 
 
+def _fix_struct_arr_names(arr, pa_type):
+    """Fix the names of the fields in a struct array to match the Arrow type.
+    This is necessary since our C++ code may not preserve the field names in
+    struct arrays.
+    """
+
+    if not pa.types.is_struct(arr.type):
+        return arr
+
+    if arr.type == pa_type:
+        return arr
+
+    if isinstance(arr, pa.ChunkedArray):
+        arr = arr.combine_chunks()
+
+    new_arrs = [
+        _fix_struct_arr_names(arr.field(i), pa_type.field(i).type)
+        for i in range(arr.type.num_fields)
+    ]
+    names = [pa_type.field(i).name for i in range(pa_type.num_fields)]
+    new_arr = pa.StructArray.from_arrays(new_arrs, names)
+    # Arrow's from_arrays ignores nulls (bug as of Arrow 19) so we add them back
+    # manually
+    return pa.Array.from_buffers(
+        new_arr.type, len(new_arr), arr.buffers()[:1], children=new_arrs
+    )
+
+
+def _arrow_to_pd_array(arrow_array, pa_type):
+    """Convert a PyArrow array to a pandas array with the specified Arrow type."""
+
+    # Our type inference may fail for some object columns so use the proper Arrow type
+    if pa_type == pa.null():
+        pa_type = arrow_array.type
+
+    # Our C++ code may not preserve the field names in struct arrays
+    # so we fix them here to match the Arrow schema.
+    if pa.types.is_struct(arrow_array.type):
+        arrow_array = _fix_struct_arr_names(arrow_array, pa_type)
+
+    # Cast to expected type to match Pandas (as determined by the frontend)
+    if pa_type != arrow_array.type:
+        arrow_array = arrow_array.cast(pa_type)
+
+    return pd.array(arrow_array, dtype=pd.ArrowDtype(pa_type))
+
+
+def arrow_table_to_pandas(arrow_table, arrow_schema=None):
+    """Convert a PyArrow Table to a pandas DataFrame. Not using Table.to_pandas()
+    since it doesn't use ArrowDtype and has issues (e.g. repeated column names fails).
+
+    Args:
+        arrow_table (pa.Table): The input Arrow table.
+        arrow_schema (pa.Schema, optional): The schema to use for the DataFrame.
+            If None, uses the schema from the Arrow table.
+
+    Returns:
+        pd.DataFrame: The converted pandas DataFrame.
+    """
+    if arrow_schema is None:
+        arrow_schema = arrow_table.schema
+
+    df = pd.DataFrame(
+        {
+            i: _arrow_to_pd_array(arrow_table.columns[i], field.type)
```

Comment:
The comment on 833 isn't clear to me.  How is that line handling duplicate names and how is that not the equivalent of doing "field.name:" here instead of "i:"?

Reply:
`"field.name:"` in a dictionary would replace duplicated values. Updated the comment.